src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -562,17 +682,35 @@ protected virtual void ValidateAtHash(OpenIdConnectProtocolValidationContext val
                 return;
             }
 
-            object atHashClaim;
-            if (!validationContext.ValidatedIdToken.Payload.TryGetValue(JwtRegisteredClaimNames.AtHash, out atHashClaim))
-                throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolInvalidAtHashException(LogHelper.FormatInvariant(LogMessages.IDX21312, validationContext.ValidatedIdToken)));
+            string atHash;
+            string alg;
+
+            if (validationContext.ValidatedIdTokenJwt != null)
+            {
+                if (!validationContext.ValidatedIdTokenJwt.TryGetPayloadValue(JsonWebTokens.JwtRegisteredClaimNames.AtHash, out object atHashClaim))
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolInvalidAtHashException(LogHelper.FormatInvariant(LogMessages.IDX21312, validationContext.ValidatedIdToken)));
+
+                atHash = atHashClaim as string;
+                if (atHash == null)
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolInvalidAtHashException(LogHelper.FormatInvariant(LogMessages.IDX21311, validationContext.ValidatedIdToken)));
+
+                var idToken = validationContext.ValidatedIdTokenJwt;
```

Comment:
>idToken [](http://example.com/codeflow?start=20&length=7)

null check needed?

Reply:
should be covered! see line 688, we check that validationContext.ValidatedIdTokenJwt is not null